### PR TITLE
Flax t5 Encoder

### PR DESCRIPTION
**Pull request by crystina-z**
# What does this PR do?

I notice that there hasn't been a T5EncoderModel for Flax implementation, taking a stab to add it. I didn't find an issue related to this.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

In terms of the test, I wrote it under `pytest tests/models/t5/test_modeling_flax_t5.py` now but seems I'm having trouble running it on my machine so far, although I can import the `FlaxT5EncoderModel` externally and able to use it successfully. I'm working on that, and just want to send the PR first to know if there's more I should add.

## Who can review?  

t5: @patrickvonplaten, @patil-suraj

### CODE REVIEW

File: src/transformers/models/t5/modeling_flax_t5.py
Comment (patrickvonplaten) [inline src/transformers/models/t5/modeling_flax_t5.py:1359]: No need for a doc string here :-)

Comment (crystina-z) [reply src/transformers/models/t5/modeling_flax_t5.py]: oops added accidentally. removed

**Comment by patrickvonplaten** [comment]
Hey @crystina-z, 

The PR looks very nice already - think we're close to merging it! Great job so far :-) A FlaxT5Encoder model class is very useful (also to build Diffusion Pipelines like Imagen cc @borisdayma @patil-suraj ) 

Some tests are currently failing because the branch of the PR is not up to date with Transformers' main branch I think. 

Could you try merging the `main` branch into your PR, e.g.:

```
git pull upstream main
```

(if you called the remote `upstream`)

**Comment by patrickvonplaten** [comment]
Let me know if you need help with anything, otherwise I think we can do a final review once the Circle CI is green :-)

**Comment by crystina-z** [comment]
Hi @patrickvonplaten , thanks for comment! Lemme try merge it with main first then.

There are indeed some tests failure I'm not sure about the reason now, like `jax.errors.ConcretizationTypeError` that happens when running `tests/models/t5/test_modeling_flax_t5.py` in my local environment. might need some help later if merging with main not solves it and I can't still find the reason later :P

File: tests/models/t5/test_modeling_flax_t5.py
Comment (patil-suraj) [inline tests/models/t5/test_modeling_flax_t5.py:498]:
```suggestion
```

File: tests/models/t5/test_modeling_flax_t5.py
Comment (patil-suraj) [inline tests/models/t5/test_modeling_flax_t5.py:615]: we should just call `model(...)` here instead of `encode`
```suggestion
                    return model(input_ids=input_ids, attention_mask=attention_mask)
```

**Review by patil-suraj — COMMENTED** [review]
Thanks a lot for adding this model! I left some comments, let me know if you need help with tests.

**Comment by patil-suraj** [comment]
Hey @crystina-z 

> 1. under the run_tests_tf, it shows FAILED

You can ignore the `run_tests_tf` since it's unrelated, rebasing should fix this.

> I'm a bit confused about how should I deal with the documentation files. and if we'll need to update the documentation, do we just copy the contents under ~/tmp/test-build to docs/source/?

This means we need to add `FlaxT5EncoderModel ` and `FlaxMT5EncoderModel` in the [t5.mdx](https://github.com/huggingface/transformers/blob/main/docs/source/en/model_doc/t5.mdx) and [mt5.mdx](https://github.com/huggingface/transformers/blob/main/docs/source/en/model_doc/mt5.mdx) files respectively.

**Comment by crystina-z** [comment]
Hi @patrickvonplaten @patil-suraj All tests passed! Also updated all comments above. lmk if there's anything else to change!

**Comment by patrickvonplaten** [comment]
That's great! Reviewing now :-)

File: src/transformers/models/t5/modeling_flax_t5.py
Comment (patrickvonplaten) [inline src/transformers/models/t5/modeling_flax_t5.py:943]: ok for me!

**Review by patrickvonplaten — APPROVED** [review]
Very nice PR! Looks good to me :-) 

@patil-suraj can you do a final review?

**Review by sanchit-gandhi — APPROVED** [review]
LGTM too! Thanks for adding this module @crystina-z 👏

**Comment by crystina-z** [comment]
Nice! Thank you all for reviewing as well! Just wonder when could I expect this PR to be merged?

**Comment by sanchit-gandhi** [comment]
Thanks for being patient @crystina-z. Let's wait for @patil-suraj to give his final review then we can merge!

**Comment by patrickvonplaten** [comment]
Good to merge I think - @patil-suraj feel free to leave comments if you don't like something, but overall it's good to merge :-) 

Great job @crystina-z !!!

**Pull request merged**
